src/day6.scala

```diff
@@ -0,0 +1,81 @@
+// using scala 3.1.0
+
+import scala.util.Using
+import scala.collection.mutable
+import scala.io.Source
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day6"))(_.mkString.trim)
+
+// "Find a way to simulate lanternfish. How many lanternfish would there be after 80
+// days?"
+def part1(input: String): Int =
+  simulate(
+    days = 80,
+    initialPopulation = Fish.parseSeveral(input)
+  )
+
+// "You can model each fish as a single number that represents the number of days
+// until it creates a new lanternfish."
+case class Fish(timer: Int)
+
+object Fish:
+  // "Suppose you were given the following list:
+  //
+  // 3,4,3,1,2
+  //
+  // This list means that the first fish has an internal timer of 3, the second fish
+  // has an internal timer of 4, and so on until the fifth fish, which has an
+  // internal timer of 2."
+  def parseSeveral(input: String): Seq[Fish] =
+    for timerStr <- input.split(",")
+      yield Fish(timerStr.toInt.ensuring(_ >= 0))
+
+/**
+ * Simulate the evolution of the population and return the number
+ * of fishes at the end of the simulation.
+ * @param days Number of days to simulate
+ * @param initialPopulation Initial population
+ */
+def simulate(days: Int, initialPopulation: Seq[Fish]): Int =
+  (1 to days)
+    .foldLeft(initialPopulation)((population, _) => tick(population))
+    .size
+
+/**
+ * Compute a new population after one day passes.
+ * @param population Current population
+ * @return New population
+ */
+def tick(population: Seq[Fish]): Seq[Fish] =
+  population.flatMap { fish =>
+    // "Each day, a `0` becomes a `6` and adds a new `8` to the end of the list"
+    if fish.timer == 0 then
+      Seq(Fish(6), Fish(8))
+    // "while each other number decreases by 1"
+    else
+      Seq(Fish(fish.timer - 1))
+  }
+
+// "How many lanternfish would there be after 256 days?"
+def part2(input: String): BigInt =
+  Fish.parseSeveral(input)
+    .map(fish => descendants(256 - fish.timer))
+    .sum
+
+// Use a cache to memoize the number of fishes created by one fish for some
+// number of remaining days to live
+val cache = mutable.Map.empty[Int, BigInt]
+
+def descendants(remainingDays: Int): BigInt =
+  cache.getOrElseUpdate(
+    remainingDays,
+    if remainingDays <= 0 then 1
+    else descendants(remainingDays - 7) + descendants(remainingDays - 9)
+  )
```

Comment:
The recursive calls to `descendants` modify the `cache`, in the middle of the `getOrElseUpdate`. This basically amounts to *concurrent modifications*. While [`mutable.Map.getOrElseUpdate`](https://www.scala-lang.org/api/2.13.7/scala/collection/mutable/Map.html#getOrElseUpdate(key:K,op:=%3EV):V) says that concurrent modifications have a specified behavior, it is not good enough for those purposes:

> Concurrent map implementations may evaluate the expression op multiple times, **or may evaluate op without inserting the result**.

(emphasis mine)

I don't know if `getOrElseUpdate` can be safely used in this context. I think that, to be safe, you need to use `get`, then compute recursive calls, then `update`.

Reply:
That’s a good point. I didn’t check that.

That being said, I think the code I wrote stays correct. The spec says that it could happen that some results are not effectively cached. In our case, this does not have an impact on the correctness of the algorithm.

I tried moving the cache management outside of the recursive calls but then the performance was 10 times worse than the current implementation (143.7 seconds vs 1.3 s), so, I would just keep the current one.